### PR TITLE
EIP649 reward reduction and difficulty bomb delay

### PR DESCRIPTION
**Pull request by pirapira**
The reward-reduction and difficulty bomb delay have been both implemented.

Tests are updated so that `testeth` with no options passes.

### CODE REVIEW

File: test/tools/jsontests/BlockChainTests.cpp [hunk 1]
```diff
@@ -531,8 +552,9 @@ void testBCTest(json_spirit::mObject& _o)
 		//check the balance before and after the block according to mining rules
 		if (blockFromFields.blockHeader().parentHash() == preHash)
 		{
+			bool const isByzantiumOrLater = ChainBranch::networkIsByzantiumOrLater(blockChainName);
```
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: @winsvega what's the best way to know if it's already Byzantium here?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 2]
```diff
@@ -563,10 +585,10 @@ void testBCTest(json_spirit::mObject& _o)
 	ImportTest::compareStates(postState, blockchain.topBlock().state());
 }
 
-bigint calculateMiningReward(u256 const& _blNumber, u256 const& _unNumber1, u256 const& _unNumber2)
+bigint calculateMiningReward(u256 const& _blNumber, u256 const& _unNumber1, u256 const& _unNumber2, bool isByzantium)
 {
 	unique_ptr<SealEngineFace> se(ChainParams(genesisInfo(test::TestBlockChain::s_sealEngineNetwork)).createSealEngine());
-	bigint baseReward = se->chainParams().blockReward;
+	bigint baseReward = se->chainParams().blockReward(isByzantium);
```
Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: this should be done inside chain params. the thing is that blockReward in this case will depend on blockNumber and evm schedule like other fork dependent params.

isByzantium = number  >=  chainParams().u256Param("ByzantiumBlockNumber")

File: libethcore/ChainOperationParams.cpp [hunk 3]
```diff
@@ -66,3 +66,16 @@ u256 ChainOperationParams::u256Param(string const& _name) const
 
 	return u256(fromBigEndian<u256>(fromHex(at)));
 }
+
+u256 ChainOperationParams::blockReward(bool isByzantiumOrLater) const
+{
+	if (isByzantiumOrLater)
+		return u256(3 * (bigint(10) ^ 18)); // 3 ETH
```
Comment (chriseth) [inline libethcore/ChainOperationParams.cpp]: Is `^` implemented as exp or as xor?

Comment (chriseth) [inline libethcore/ChainOperationParams.cpp]: I think there should also be a global constant called `ether` somewhere with the right value.

Comment (pirapira) [inline libethcore/ChainOperationParams.cpp]: I wondered these too.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 2 above)
Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: Thank you.

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: I mean it feels like isByzantium should be calculated inside chainParams. 
Cause it knows the schedule from s_sealEngineNetwork. if should be able to return the reward if you give it the blocknumber

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: I'll try.

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: No, `ChainParams` does not know the schedule.  It does not have access to a `sealEngineNetwork`.

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: it has the genesis transition block number for sure. 
how is it done for schedule variables?

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: I mean it is strange that we calculate isBizantium manually outside of getBlockReward

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: What is the genesis transition block?  What are schedule variables?

Is `ChainParams` supposed to know when Byzantium starts?  In that case, I will need to fix `ChainParams` class first.

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: ChainParams  constructed from genesisInfo so it does know all the transition blocks. 
Schedule variables are: EVMSchedule.h

Ah, and here there is a function 
EVMSchedule const& SealEngineBase::evmSchedule(u256 const& _blockNumber) const

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: OK.  I'll file another PR first to fix this first.

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: @winsvega so, the SealEngine knows when Byzantium starts.  ChainParams should not know that.

Comment (winsvega) [inline test/tools/jsontests/BlockChainTests.cpp]: why? chainParams initialized from genesisInfo. it should know that. cause seal engine knows that from chainParams.

``` chainParams().u256Param("byzantiumForkBlock")  ```  
indicates when Byzanium starts.  

do you think we should look at this issue more carefully and perhaps do a refactoring PR for libethereum instead?

Comment (pirapira) [inline test/tools/jsontests/BlockChainTests.cpp]: No, I came up with one solution.

**Comment by pirapira** [comment]
@winsvega shall these cases appear in the test case spreadsheet?

**Comment by winsvega** [comment]
I mean we could add the tests with this PR. that was the idea of a submodule. 
to cover the PR changes with relevant tests

**Comment by pirapira** [comment]
@winsvega I prefer a separate PR for adding new tests.  We should deliver the fixed tests earlier, even without the new tests.

**Comment by winsvega** [comment]
if we have the tests at least we know the change on this PR is covered. that what is the point

**Comment by pirapira** [comment]
Let me see if I need to fix any fillers.  If yes, we have the changes covered somehow.

**Comment by pirapira** [comment]
I had an `ethereum/tests` commit on my local machine, pushed that and used that in the submodule.

**Comment by chfast** [comment]
At first look, I don't like how we manage chain configuration. Do we need to use JSON-like structure and constructs like "get u256 param" all the way to the bottom?

**Comment by pirapira** [comment]
@chfast I believe we should do
```
_bc.chainParams().EIP158ForkBlock()
```
instead of
```
_bc.chainParams().u256Param("EIP158ForkBlock")
```
And, remove `u256Param` from the public interface of `ChainParams`.

I want to do it in a separate PR.  Do you want it before this one?

**Comment by chfast** [comment]
After.

I'd like to review the chain config design later on...
For EVM-C single enum is enough. Also we can consider configs for chains + separated config for chain switches (hard forks).

File: libethcore/EVMSchedule.h [hunk 4]
```diff
@@ -71,6 +72,8 @@ struct EVMSchedule
 	unsigned blockhashGas = 20;
 	unsigned maxCodeSize = unsigned(-1);
 
+	boost::optional<u256> blockRewardOverwrite{};
```
Comment (chfast) [inline libethcore/EVMSchedule.h]: Explicit initialization not needed.

File: libethcore/ChainOperationParams.h [hunk 5]
```diff
@@ -73,7 +73,11 @@ struct ChainOperationParams
 	std::string sealEngineName = "NoProof";
 
 	/// General chain params.
-	u256 blockReward = 0;
+private:
+	u256 m_blockReward = 0;
```
Comment (chfast) [inline libethcore/ChainOperationParams.h]: Explicit initialization not needed.

**Comment by pirapira** [comment]
@chfast removed the unnecessary initializations.

**Pull request merged**